src/Evolution/Systems/ScalarWave/Constraints.cpp

```diff
@@ -0,0 +1,98 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/ScalarWave/Constraints.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace ScalarWave {
+template <size_t Dim>
+tnsr::i<DataVector, Dim, Frame::Inertial> one_index_constraint(
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& d_psi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi) noexcept {
+  auto constraint =
+      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(phi, 0.);
+  one_index_constraint<Dim>(make_not_null(&constraint), d_psi, phi);
+  return constraint;
+}
+
+template <size_t Dim>
+void one_index_constraint(
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*> constraint,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& d_psi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi) noexcept {
+  if (get_size(get<0>(*constraint)) != get_size(get<0>(phi))) {
+    *constraint =
+        make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(phi, 0.);
+  }
+  // Declare iterators for d_psi and phi outside the for loop,
+  // because they are const but constraint is not
+  auto d_psi_it = d_psi.begin(), phi_it = phi.begin();
+
+  for (auto constraint_it = (*constraint).begin();
+       constraint_it != (*constraint).end();
+       ++constraint_it, (void)++d_psi_it, (void)++phi_it) {
+    *constraint_it = *d_psi_it - *phi_it;
+  }
+}
+
+template <size_t Dim>
+tnsr::ij<DataVector, Dim, Frame::Inertial> two_index_constraint(
+    const tnsr::ij<DataVector, Dim, Frame::Inertial>& d_phi) noexcept {
+  auto constraint =
+      make_with_value<tnsr::ij<DataVector, Dim, Frame::Inertial>>(d_phi, 0.);
+  two_index_constraint<Dim>(make_not_null(&constraint), d_phi);
+  return constraint;
+}
+
+template <size_t Dim>
+void two_index_constraint(
+    const gsl::not_null<tnsr::ij<DataVector, Dim, Frame::Inertial>*> constraint,
+    const tnsr::ij<DataVector, Dim, Frame::Inertial>& d_phi) noexcept {
+  if (get_size(get<0, 0>(*constraint)) != get_size(get<0, 0>(d_phi))) {
+    *constraint =
+        make_with_value<tnsr::ij<DataVector, Dim, Frame::Inertial>>(d_phi, 0.);
+  }
+  // Not using the antisymmetric property of the 2-index constraint, via
+  // LeviCivita iterators, here as for Dim = 1 or 2 the constraint then becomes
+  // a scalar.
```

Comment:
Why not just make the tag hold an antisymmetric tensor?

Reply:
Not sure what you mean here. What my comment means is that it is possible to redefine the 2-index constraint as `C_{ij} --> \eps_{k}^{ij} C_{ij} --> D_{k}` as a 1-index constraint, but only in 3 and higher spatial dimensions. If we were working in 3+ dimensions, we could use this simplification to save some memory allocations, but we cannot do that here since we have to generalize to 1 and 2D.

Reply:
I realize that that comment of mine was adding little value, as it was about something I'm not doing. So I've removed it, removing any confusion.

Reply:
@nilsleiffischer we don't have support for antisymmetric tensors.